src/sysinfo.cc

```diff
@@ -544,6 +553,80 @@ int GetNumCPUs() {
   BENCHMARK_UNREACHABLE();
 }
 
+class ThreadAffinityGuard final {
+ public:
+  ThreadAffinityGuard() {
+    reset_affinity = [this] {
+#if defined(BENCHMARK_HAS_PTHREAD_AFFINITY)
+      int ret;
+      self = pthread_self();
+      ret = pthread_getaffinity_np(self, sizeof(previous_affinity),
+                                   &previous_affinity);
+      if (ret != 0) return false;
+
+      cpu_set_t affinity;
+      memcpy(&affinity, &previous_affinity, sizeof(affinity));
+
+      bool is_first_cpu = true;
+
+      for (int i = 0; i < CPU_SETSIZE; ++i)
+        if (CPU_ISSET(i, &affinity)) {
+          if (is_first_cpu)
+            is_first_cpu = false;
+          else
+            CPU_CLR(i, &affinity);
+        }
+
+      if (is_first_cpu) return false;
+
+      ret = pthread_setaffinity_np(self, sizeof(affinity), &affinity);
+      return ret == 0;
+#elif defined(BENCHMARK_OS_WINDOWS_WIN32)
+      self = GetCurrentThread();
+      DWORD_PTR mask = static_cast<DWORD_PTR>(1) << GetCurrentProcessorNumber();
+      previous_affinity = SetThreadAffinityMask(self, mask);
+      return previous_affinity != 0;
+#else
+      return false;
+#endif
```

Comment:
`#endif  // BENCHMARK_HAS_PTHREAD_AFFINITY`

or similar please, to make this easier to reason about.